Read the Table_Character correctly to read to the end of the record b…

qchaupds opened the pull request:
…ased on record length instead of the normal <CR><LF> and null pointer should not show up.

Closes NASA-PDS/validate#325, NASA-PDS/validate#335

This pull request resolves two issues: Read the Table_Character correctly and corrected the null pointer when checking for a command line parameter.

Tested in DEV for #325
Previously would result in ERROR because it was reading the file line by line, which causes the record to end prematurely

% validate -R pds4.label --skip-context-validation -r report_github325_label_valid.json -s json -t src/test/resources/github325/crs009x.xml

Tested in DEV for #335
Three different ways to process a target, none should throw an exception.

% validate -R pds4.label  --skip-context-validation -r report_github335_as_label.json   -s json -t src/test/resources/github335/minimal_test_product.xml 
% validate   --skip-context-validation -r report_github335_as_target.json   -s json -t src/test/resources/github335/*.xml 
% validate   --skip-context-validation -r report_github335_as_dir.json   -s json -t src/test/resources/github335

Commit:
Read the Table_Character correctly to read to the end of the record based on record length instead of the normal <CR><LF> and null pointer should not show up.

1.  Correct the null pointer issue in RuleContext.java
2.  Move part of a long function into a shorter function validateTableContentLineWise() in TableDataContentValidationRule.java
3.  Add a peek into a text label for length of first line in TableDataContentValidationRule.java
4.  Add new function validateTableContentRecordWise() to process one record at a time according to record_length in TableDataContentValidationRule.java

Refs:

https://github.com/NASA-PDS/validate/issues/325 Validate Incorrectly Throws Error When Embedded Field_Character Contains <CR><LF>
https://github.com/NASA-PDS/validate/issues/335 validate gives a NullPointerException during validation of a directory containing Table_Character products

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -166,6 +168,339 @@ public boolean isApplicable(String location) {
+  }
+
+  private boolean isTableLineOriented(Object table, URL dataFile) {
+      // Given a text table object, determine if it a line oriented table or not.  A line-oriented table contains some record delimiter and optional record_length.
```
@qchaupds can we add this as a javadoc?

```java
/**
 * Given a text table object, determine if it a line oriented table or not.  A l ...
 **/
```

qchaupds replied:
Done now and pushed to github.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -166,6 +168,339 @@ public boolean isApplicable(String location) {
+
+
+  private long getFirstLineLength(URL dataFile) {
+      // Take a peek and read the first line from the file and retrieve the line length.
```
@qchaupds can we add this as a javadoc?

```java
/**
 * Given a text table object, determine if it a line oriented table or not.  A l ...
 **/
```

qchaupds replied:
Done now and pushed to github.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -166,6 +168,339 @@ public boolean isApplicable(String location) {
+          }
+          // Need to manually parse the line if we're past the defined
+          // number of records. The PDS4-Tools library won't let us
+          // read past the defined number of records.
```
@qchaupds if there is data beyond the number of records defined, this can be considered a "footer" for the data and we should not assume anything about this data and/or manually read the data.

we should just throw a warning that there is more data than expected based upon the number of rows (I thought we already did this?).

this could be used for calculating the file size, but we should not do any more validation with the assumption these are actual rows of the table

jordanpadams replied:
looking at the code below, i may be misunderstanding what is happening here, but we shouldn't be checking any of the data in the "footer" of the data. we should just throw a warning and use the extra info to calculate the file size

qchaupds replied:
The code you are asking about is extracted AS-IS from a very large function.   Perhaps we can look at this in a separate ticket.

Commit:
Add javadoc to several functions

jordanpadams reviewed (changes requested):
@qchaupds looks like we are missing the test data

Commit:
Add missing test resources

jordanpadams committed:
Update to latest PDS4 Jparser SNAPSHOT

jordanpadams committed:
fix for backwards compatibility with java8

cast ByteBuffer to Buffer

jordanpadams committed:
Add common error to docs for backwards incompatibility

when trying to run validate with bundle validation against a directory in java8, the software just hangs. updated docs with a workaround. refs #336

jordanpadams committed:
Revert "Add common error to docs for backwards incompatibility"

This reverts commit a0158b81e636cbb61bb15e6996f1759e75f994b4.

jordanpadams committed:
Update regex to be more valid

This seems to work fine in newer versions of java, but java8 just hangs and doesn't know what do with those extra wildcards.

resolves #336
